addons/sale/models/product_template.py

```diff
@@ -118,3 +118,24 @@ def get_import_templates(self):
                     'template': '/product/static/xls/product_template.xls'
                 }]
         return res
+
+    @api.model
+    def _get_incompatible_types(self):
+        return []
+
+    def _check_incompatible_types(self):
+        incomptatible_types = self._get_incompatible_types()
+        if len(incomptatible_types) < 2:
+            return
+        fields = self.env['ir.model.fields'].sudo().search_read(
+            [('model', '=', 'product.template'), ('name', 'in', incomptatible_types)],
+            ['field_description'])
+        field_descriptions = [v['field_description'] for v in fields]
+        field_list = incomptatible_types + ['name']
+        values = self.read(field_list)
+        for val in values:
+            incompatible_values = [val[f] for f in self._get_incompatible_types() if val[f]]
+            if len(incompatible_values) > 1:
+                raise ValidationError(_("The product (%s) has incompatible values: %s",
+                                        val['name'], ','.join(field_descriptions)))
+        return
```

Comment:
you don't need return without value as last line of a method, this is equivalent to not having "return"

(same for the two enterprise override)